numba_dpex/core/utils/kernel_launcher.py

```diff
@@ -374,78 +374,94 @@ def get_queue(self, exec_queue: dpctl.SyclQueue) -> llvmir.Instruction:
         )
         return self.builder.load(sycl_queue_val)
 
-    def _allocate_kernel_arg_array(self, num_kernel_args):
-        """Allocates an array to store the LLVM Value for every kernel argument.
+    def _allocate_array(
+        self, numba_type: types.Type, size: int
+    ) -> llvmir.Instruction:
+        """Allocates an LLVM array of given type and size.
 
         Args:
-            num_kernel_args (int): The number of kernel arguments that
-            determines the size of args array to allocate.
+            numba_type: type of the array to allocate,
+            size: The size of args array to allocate.
 
-        Returns: An LLVM IR value pointing to an array to store the kernel
-        arguments.
+        Returns: An LLVM IR value pointing to the array.
         """
-        args_list = cgutils.alloca_once(
+        return cgutils.alloca_once(
             self.builder,
-            utils.LLVMTypes.byte_ptr_t,
-            size=self.context.get_constant(types.uintp, num_kernel_args),
+            self.context.get_value_type(numba_type),
+            size=self.context.get_constant(types.uintp, size),
         )
 
-        return args_list
+    def _populate_array_from_python_list(
+        self,
+        numba_type: types.Type,
+        py_array: list[llvmir.Instruction],
+        ll_array: llvmir.Instruction,
+        force_cast: bool = False,
+    ):
+        """Populates llvmir values from the input array into newly create array.
 
-    def _allocate_kernel_arg_ty_array(self, num_kernel_args):
-        """Allocates an array to store the LLVM Value for the typenum for
-        every kernel argument.
+        Args:
+            numba_type: type of the array to allocate,
+            py_array: array of llvm ir values to populate.
+            ll_array: llvm ir value that represents an array to populate,
+            force_cast: either force cast values to the provided type.
+        """
+        for idx, ll_value in enumerate(py_array):
+            ll_array_dst = self.builder.gep(
+                ll_array,
+                [self.context.get_constant(types.int32, idx)],
+            )
+            # bitcast may be extra, but won't hurt,
+            if force_cast:
+                ll_value = self.builder.bitcast(
+                    ll_value,
+                    self.context.get_value_type(numba_type),
+                )
+            self.builder.store(ll_value, ll_array_dst)
+
+    def _create_ll_from_py_list(
+        self,
+        numba_type: types.Type,
+        array: list[llvmir.Instruction],
+        force_cast: bool = False,
+    ) -> llvmir.Instruction:
+        """Allocates llvm array same size as input python array. Then it
```

Comment:
Allocates an LLVM IR array of the same size as the input python list of LLVM IR Values and populates the array with the LLVM Values in the list.